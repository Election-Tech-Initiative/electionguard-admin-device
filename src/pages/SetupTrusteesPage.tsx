import React, { useContext, PointerEventHandler } from 'react'
import styled from 'styled-components'
import ElectionContext from '../contexts/electionContext'
import Main, { MainChild } from '../components/Main'
import Button, { SegmentedButton } from '../components/Button'
import TextIcon from '../components/TextIcon'
import Screen from '../components/Screen'
import LinkButton from '../components/LinkButton'
import ElectionInfo from '../components/ElectionInfo'
import Sidebar from '../components/Sidebar'
import Prose from '../components/Prose'
import { ValidTrusteeCount } from '../config/types'
import { MAX_TRUSTEES } from '../config/globals'

const Header = styled.div`
  margin: 0 auto;
  text-align: center;
`

const LogoImage = styled.img`
  display: flex;
  margin: 0 auto;
  max-width: 12rem;
`

const SetupTrusteesPage = () => {
  const {
    election,
    electionGuardConfig,
    setNumberOfTrustees,
    setThreshold,
  } = useContext(ElectionContext)
  const { numberOfTrustees, threshold } = electionGuardConfig

  if (!threshold && !numberOfTrustees) {
    setNumberOfTrustees(1)
    setThreshold(1)
  }

  const onNumberOfTrusteesChange: PointerEventHandler = event => {
    const target = event.target as HTMLInputElement
    const value = +target.value as ValidTrusteeCount
    if (value < electionGuardConfig.threshold) {
      setThreshold(value)
    }
    setNumberOfTrustees(value)
  }

  const onThresholdChange: PointerEventHandler = event => {
    const target = event.target as HTMLInputElement
    const value = +target.value as ValidTrusteeCount
    if (value > numberOfTrustees) {
      setThreshold(numberOfTrustees)
    } else {
      setThreshold(value)
    }
  }

  const trusteeOptions = Array.from(Array(MAX_TRUSTEES).keys()).map(x => x + 1)
  const thresholdOptions = Array.from(Array(numberOfTrustees).keys()).map(
    x => x + 1
  )

  return (
    <Screen>
      <Main>
        <MainChild>
          <Prose id="audiofocus">
            <Header>
              <h1>Setup Trustees</h1>
            </Header>
          </Prose>
          <Prose textCenter>
            <p
              aria-label={`Select number of trustees available. You have selected ${numberOfTrustees} trustee(s). Use the down arrow to hear your options. Use the right arrow to move to the next contest.`}
            >
              <b>How many trustees available? </b>
            </p>
            <p> {numberOfTrustees} trustee(s) </p>
            <SegmentedButton data-testid="change-number-of-trustee-buttons">
              {trusteeOptions.map((value: number, index: number) => (
                <Button
                  key={value}
                  data-size={index}
                  small
                  onPress={onNumberOfTrusteesChange}
                  value={value}
                  primary={value === numberOfTrustees}
                >
                  {value}
                </Button>
              ))}
            </SegmentedButton>
            <p
              aria-label={`Select number of trustees required to decrypt. You have selected ${threshold} trustee(s). Use the down arrow to hear your options. Use the right arrow to move to the next contest.`}
            >
              <b>How many trustees required to decrypt?</b>
            </p>
            <p> {threshold} trustee(s) </p>
            <SegmentedButton data-testid="change-threshold-buttons">
              {thresholdOptions.map((value: number, index: number) => (
                <Button
                  key={value}
                  data-size={index}
                  small
                  onPress={onThresholdChange}
                  value={value}
                  primary={value === threshold}
                >
                  {value}
                </Button>
              ))}
            </SegmentedButton>
          </Prose>
        </MainChild>
      </Main>
      <Sidebar
        footer={
          <>
            <hr />
            <ElectionInfo election={election} precinctId="" horizontal />
            <hr />
            <LogoImage
              alt="Election Guard Logo"
              src="/images/electionguard.svg"
            />
          </>
        }
      >
        <p>
          <LinkButton big primary to="/keys" id="next">
            <TextIcon arrowRight white>
              Next
            </TextIcon>
          </LinkButton>
        </p>
        <p>
          <LinkButton small to="/start" id="back">
            <TextIcon arrowLeft>Back</TextIcon>
          </LinkButton>
        </p>
      </Sidebar>
    </Screen>
  )
}

export default SetupTrusteesPage
