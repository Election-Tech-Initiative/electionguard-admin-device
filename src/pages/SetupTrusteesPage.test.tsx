import React from 'react'
import { Route } from 'react-router-dom'

import { render } from '../../test/testUtils'

import SetupTrusteesPage from './SetupTrusteesPage'

it('renders SetupElectionPage', async () => {
  const { container } = render(
    <Route path="/" component={SetupTrusteesPage} />,
    {
      route: '/',
    }
  )
  expect(container.firstChild).toMatchSnapshot()
})
