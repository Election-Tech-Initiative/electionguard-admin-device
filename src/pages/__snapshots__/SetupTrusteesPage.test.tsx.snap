// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders SetupElectionPage 1`] = `
<div
  class="sc-bZQynM ebsnPk"
>
  <main
    class="sc-bdVaJa iiFDUH"
  >
    <div
      class="sc-bwzfXH cJzayE"
    >
      <div
        class="sc-dnqmqq bznNpt"
        id="audiofocus"
      >
        <div
          class="sc-chPdSV dAbUGJ"
        >
          <h1>
            Setup Trustees
          </h1>
        </div>
      </div>
      <div
        class="sc-dnqmqq buYLZL"
      >
        <p
          aria-label="Select number of trustees available. You have selected undefined trustee(s). Use the down arrow to hear your options. Use the right arrow to move to the next contest."
        >
          <b>
            How many trustees available? 
          </b>
        </p>
        <p>
           
           trustee(s) 
        </p>
        <span
          class="sc-ifAKCX eqnyTq"
          data-testid="change-number-of-trustee-buttons"
        >
          <button
            class="sc-bxivhb cOyoGh"
            data-size="0"
            type="button"
            value="1"
          >
            1
          </button>
          <button
            class="sc-bxivhb cOyoGh"
            data-size="1"
            type="button"
            value="2"
          >
            2
          </button>
          <button
            class="sc-bxivhb cOyoGh"
            data-size="2"
            type="button"
            value="3"
          >
            3
          </button>
          <button
            class="sc-bxivhb cOyoGh"
            data-size="3"
            type="button"
            value="4"
          >
            4
          </button>
          <button
            class="sc-bxivhb cOyoGh"
            data-size="4"
            type="button"
            value="5"
          >
            5
          </button>
        </span>
        <p
          aria-label="Select number of trustees required to decrypt. You have selected undefined trustee(s). Use the down arrow to hear your options. Use the right arrow to move to the next contest."
        >
          <b>
            How many trustees required to decrypt?
          </b>
        </p>
        <p>
           
           trustee(s) 
        </p>
        <span
          class="sc-ifAKCX eqnyTq"
          data-testid="change-threshold-buttons"
        >
          <button
            class="sc-bxivhb cOyoGh"
            data-size="0"
            type="button"
            value="1"
          >
            1
          </button>
        </span>
      </div>
    </div>
  </main>
  <nav
    class="sc-fjdhpX esZIZf"
  >
    <div
      class="sc-cSHVUG eLCTxm"
    >
      <p>
        <button
          class="sc-bxivhb rdECE"
          id="next"
          role="option"
          type="button"
        >
          <span
            class="sc-EHOje bZxfGe"
          >
            Next
          </span>
        </button>
      </p>
      <p>
        <button
          class="sc-bxivhb cOyoGh"
          id="back"
          role="option"
          type="button"
        >
          <span
            class="sc-EHOje zoDGT"
          >
            Back
          </span>
        </button>
      </p>
    </div>
    <div
      class="sc-kAzzGY gObJYC"
    >
      <hr />
      <div
        aria-hidden="true"
        class="sc-VigVT gbgFAa"
        data-testid="election-info"
      >
        <div
          class="sc-jTzLTM gIFeXt"
        >
          <div
            aria-hidden="true"
            class="sc-gzVnrw kgzhsX"
          />
          <div
            class="sc-dnqmqq eppipW"
          >
            <h5
              aria-label=" General Election."
            >
               General Election
            </h5>
            <p
              class="sc-iwsKbI hNZFgJ"
            >
              November 3, 2020
              <br />
              <span
                class="sc-gZMcBi hbCako"
              >
                Franklin County
                ,
              </span>
               
              <span
                class="sc-gZMcBi hbCako"
              >
                State of Hamilton
              </span>
            </p>
            <p
              class="sc-iwsKbI lhbfNZ"
            >
               
            </p>
          </div>
        </div>
      </div>
      <hr />
      <img
        alt="Election Guard Logo"
        class="sc-kgoBCf kCLXgh"
        src="/images/electionguard.svg"
      />
    </div>
  </nav>
</div>
`;
